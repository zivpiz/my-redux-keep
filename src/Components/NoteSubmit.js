import React, {Component} from 'react';
import {connect} from 'react-redux';
import {createCheckedNote, createUncheckedNote} from "../actions";

class NoteSubmit extends Component {
    constructor(props) {
        super(props);
        this.state = {text: '', checked: false};

        this.ConditionalBoxRender = this.ConditionalBoxRender.bind(this);
        this.handleTextChange = this.handleTextChange.bind(this);
        this.handleCheck = this.handleCheck.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.submitNote = this.submitNote.bind(this);
    }

    ConditionalBoxRender = () => this.state.text === '' ? '' : <input type="checkbox" onChange={this.handleCheck}/>;

    handleTextChange = (event) => this.setState({text: event.target.value});

    handleCheck = () => this.setState({checked: !this.state.checked}, () => this.submitNote());

    handleSubmit = (event) => {
        event.preventDefault();
        this.submitNote();
    };

    submitNote = () => {
        if (this.state.text === '')
            return;
        if (this.state.checked)
            this.props.createCheckedNote(Object.assign({id: this.props.nextId}, this.state));
        else
            this.props.createUncheckedNote(Object.assign({id: this.props.nextId}, this.state));
        this.setState({text: '', checked: false});
    };

    render() {
        return (
            <form onSubmit={this.handleSubmit}>
                {this.ConditionalBoxRender()}
                <input type="text" placeholder="Take a note..." name="inputName"
                       value={this.state.text} onChange={this.handleTextChange} autoComplete="off"/>
            </form>
        )
    }
}

const mapStateToProps = (state) => ({nextId: state.nextId});
export default connect(mapStateToProps, {createCheckedNote, createUncheckedNote})(NoteSubmit);