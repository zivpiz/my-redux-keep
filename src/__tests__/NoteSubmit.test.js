import NoteSubmit from "../Components/NoteSubmit";
import React from 'react';

describe("Test NoteSubmit Component ", ()=>{

    let wrapper;
    const mockUpdateNote = jest.fn();
    const mockChangeEvent = { target: { value: "Some text"}};
    const mockSubmitEvent = { preventDefault: ()=> {}};
    beforeEach(()=>{
        wrapper = shallow(<NoteSubmit updateNoteList={mockUpdateNote}/>);
    });

    test('Renders appropriate snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Doesn’t render checkbox when there’s no text', () => {
        expect(wrapper.find('input')).toHaveLength(1);
    });

    test('Changes text appropriately', () => {
        wrapper.find('input').at(0).simulate('change', mockChangeEvent);
        expect(wrapper.find('input').at(1).props().value).toBe("Some text");
    });

    test('Renders checkbox when there’s text', () => {
        wrapper.find('input').at(0).simulate('change', mockChangeEvent);
        expect(wrapper.find('input')).toHaveLength(2);
    });

    test('Calls update function when submitting form', () => {
        wrapper.find('input').at(0).simulate('change', mockChangeEvent);
        wrapper.find('form').simulate('submit', mockSubmitEvent);
        expect(mockUpdateNote).toHaveBeenCalled();
    });

    test('Calls update function when checking checkbox', () => {
        wrapper.find('input').at(0).simulate('change', mockChangeEvent); // Text Change
        wrapper.find('input').at(0).simulate('change', mockChangeEvent); // Checkbox change
        expect(mockUpdateNote).toHaveBeenCalled();
    });

    test('Resets state when checking checkbox / submitting', () => {
        wrapper.find('input').at(0).simulate('change', mockChangeEvent);
        wrapper.find('form').simulate('submit', mockSubmitEvent);
        expect(wrapper.state().text).toBe('');
        expect(wrapper.state().checked).toBe(false);
    });

    test('Blocks update function when text is empty', () => {
        wrapper.find('form').simulate('submit', mockSubmitEvent);
        expect(mockUpdateNote).toHaveBeenCalledTimes(0);
    });

    afterEach(()=>{
        mockUpdateNote.mockRestore();
    })

});


