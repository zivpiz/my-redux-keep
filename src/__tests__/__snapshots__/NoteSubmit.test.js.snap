// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test NoteSubmit Component  Renders appropriate snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    autoComplete="off"
    name="inputName"
    onChange={[Function]}
    placeholder="Take a note..."
    type="text"
    value=""
  />
</form>
`;
